### PR TITLE
Laplace distribution.  

### LINKED ISSUE #22: Laplace Distribution
Wikipedia's definition looks pretty straightforward.  I started work by copying the exponential.rs and just replacing formulas.

What's the scope of work that needs to be done on all new distributions?

### PR DESCRIPTION
**Pull request by knappador**
Mostly copied over from exponential.rs, which shares some constants and imports.

I used symbolic manipulation to derive the inverse CDF.  Others are just straight imports from wikipedia.

https://en.wikipedia.org/wiki/Laplace_distribution

closes #22

### CODE REVIEW

File: src/distribution/laplace.rs [hunk 1]
```diff
@@ -0,0 +1,265 @@
+use distribution;
+use distribution::Inverse;
+use source::Source;
+
+/// An exponential distribution.
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: Please update this one.

File: src/distribution/laplace.rs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+#[derive(Clone, Copy, Debug)]
+pub struct Laplace {
+    mu: f64,
+    b: f64,
+    _1_over_2b: f64,
+}
+
+impl Laplace {
+    /// Create a laplace distribution with location `mu` and scale `b`.
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: Please spell the name with a capital letter.

File: src/distribution/laplace.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    ///
+    /// It should hold that `b > 0`.
+    #[inline]
+    pub fn new(mu: f64, b: f64) -> Self {
+        should!(b > 0.0);
+        let _1_over_2b = b.recip() * 0.5;
+        Laplace { mu, b, _1_over_2b }
+    }
+
+    // Return mu, the location parameter
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: “Return the location parameter.”

File: src/distribution/laplace.rs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    #[inline(always)]
+    pub fn mu(&self) -> f64 {
+        self.mu
+    }
+
+    // Return b, the scale parameter
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: “Return the scale parameter.”

File: src/distribution/laplace.rs [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        (std::f64::consts::E * 2.0 * self.b).ln()
+    }
+}
+
+impl distribution::Inverse for Laplace {
+    #[inline]
+    fn inverse(&self, p: f64) -> f64 {
+        should!(0.0 <= p && p <= 1.0);
+        if p > 0.5 {
+            // p == 1.0 -> division by zero
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: Please remove.

File: src/distribution/laplace.rs [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            if p == 1.0 {
+                return std::f64::INFINITY;
+            }
+            self.mu + self.b * (-0.5 / (p - 1.0)).ln()
+        } else {
+            // p == 0.0 -> natural log of zero
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: Please remove.

File: src/distribution/laplace.rs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        let x = vec![
+            std::f64::NEG_INFINITY,
+            -2.8283137373023006,
+            -0.07944154167983575,
+            2.0,
+            4.079441541679836,
+            6.8283137373023015,
+            std::f64::INFINITY,
+        ];
+        let p = vec![0.0, 0.1, 0.25, 0.5, 0.75, 0.9, 1.000000000000000e-00];
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: What is up with the last one in the list?

File: src/distribution/laplace.rs [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    }
+
+    #[test]
+    fn variance() {
+        assert_eq!(new!(2.0, 3.0).variance(), 18.0);
+    }
+
+    #[test]
+    fn deviation() {
+        assert_eq!(new!(2.0, 3.0).deviation(), 4.242640687119286);
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: Not sure about `assert_eq!` here. Maybe `assert::close`?

File: src/distribution/laplace.rs [hunk 9]
```diff
@@ -0,0 +1,265 @@
+use distribution;
+use distribution::Inverse;
+use source::Source;
+
+/// An exponential distribution.
+#[derive(Clone, Copy, Debug)]
+pub struct Laplace {
+    mu: f64,
+    b: f64,
+    _1_over_2b: f64,
```
Comment (IvanUkhov) [inline src/distribution/laplace.rs]: Let’s inline it and remove from here.

**Comment by knappador** [comment]
I did a pretty lazy pass to address only comments. 

I replaced the expression for the quantile function for x > 0.5.  Both forms passed tests, but while checking the result by hand, I wound up with the same expression as the reference this time around:

https://en.wikipedia.org/wiki/Laplace_distribution

**Comment by IvanUkhov** [comment]
Thank you for making the adjustments!

**Pull request merged**